No instance eval

johnnymo87 opened the pull request:
@tansaku, as you mentioned in [continues refactor of base org and org code](https://github.com/AgileVentures/LocalSupport/pull/180), you were surprised by the use of `instance_eval`. Here's a PR to remove that.

tansaku commented:
thanks @johnnymo87 much appreciated.  Looks like we have some outstanding conflicts on this branch.  

Also, reviewing the code I was looking at:

```ruby
@organisation = Organisation.find(params[:id])
@markers = build_map_markers(Organisation.where(id: @organisation))
```

in organisations_controller.rb (and some others) and thinking do we need to query the org twice?  can we just do:

```ruby
@organisation = Organisation.find(params[:id])
@markers = build_map_markers(@organisation)
```

Finally, do you think the tests in proposed_organisations_controller spec are necessary.  Following the Makers philosophy I've been leaning towards only unit and feature tests, and only testing the controllers with request specs if there are security end points that are difficult to test otherwise.   Is proposed_organisations_controller_spec creating coverage for us that isn't handled by any other test?

johnnymo87 commented:
@tansaku I've rebased over develop and I've remove the duplicate queries.

Regarding the controller tests, I can understand where you're coming from with wanting to avoid controller specs. I agree that we could go without them if we're sure our cucumber coverage is sufficient and our controllers are skinny enough. I think, though, that some of our controllers are perhaps still a bit logic-heavy to go totally controller-test-free. Take `Organisations#show`, for an example. We have eight instance variables being created here. If instead we had them all tucked into a presenter, then we'd only have one line of code where the controller fetches the presenter for the view. We could go without the controller specs because we'd have specs for the presenter instead.

Either way, what we've got now are these controller specs that are mocking waaay to much. And from what @mtc2013 told me, they're the main reason he went with `instance_eval` in the first place -- cuz they're so damn brittle that you have to rewrite them every time you want to refactor the controller. So I just replaced the mocks with real model instances where necessary, and now we can refactor with ease. But if you think they're unneeded altogether, LMK and I'll just delete them entirely.

tansaku commented on app/controllers/organisations_controller.rb:
```diff
@@ -27,15 +27,15 @@ def index
     @can_propose_edits = current_user.present? && !@editable
-   # @next_path = current_user ? organisation_user_path(@organisation.id, current_user.id) : new_user_session_path
-    @markers = build_map_markers([@organisation])
+    @markers = build_map_markers(organisations)
```
can we not just get the first organisation with find, and then pass [@organisation] to build_map_markers?

tansaku commented:
thanks @johnnymo87 I totally hear you as regards the specs.  Ultimately the goal should be to pull as much logic as possible out of the controllers and into models, concerns, services and what have you.

I'm not too bothered if there are controller specs hanging around, but when there are new ones the question I want to ask is does this new spec cover a code path that isn't covered by the corresponding feature test, and/or is it a smell that's pushing us to move some more logic out of the controller.

I've been using simplecov and coveralls a lot recently to track code paths.  I guess we're not set up to get easy stats on that - perhaps I should set up coveralls on localsupport so we can all see those things more easily ...

johnnymo87 commented:
OK, I see that `features/individual_charity_pages/propose_new_organisation.feature` covers the show action that I was testing in the new controller spec, so I've removed that spec file.

tansaku commented:
thanks @johnnymo87 any thoughts on that inline comment I made about the controller actions:

> can we not just get the first organisation with find, and then pass [@organisation] to build_map_markers?

```
  def show
    organisations = Organisation.where(id: params[:id])
    @organisation = organisations.first!
   ...
    @markers = build_map_markers(organisations)
  end
```

could be: 

```
  def show
    @organisation = Organisation.find(params[:id])
   ...
    @markers = build_map_markers([@organisation])
  end
```

mtc2013 commented:
@tansaku @johnnymo87 unfortunately it needs to be a scope rather than an array

johnnymo87 commented:
Yeah, we're using an ActiveRecord `select` [in the query](https://github.com/johnnymo87/LocalSupport/blob/no-instance-eval/app/services/queries/organisations.rb#L32), not a ruby enumerable `select`.

tansaku commented:
fair enough :-)
